contributors/design-proposals/pod-priority-resourcequota.md

```diff
@@ -0,0 +1,74 @@
+Pods can now be classified into different priority classes. We would like to be able to create quota for various priority classes in order to manage cluster resources better and limit abuse scenarios. In order to implement this, we need to add a priority class name filed to ResourceQuota definition.
+
+## Overview 
+
+This design doc introduces a new field in ResourceQuota to specify a priority class and explains how quota enforcement logic is changed to apply the quota to pods with the given priority class.
+
+## Detailed Design 
+
+### Expected behavior of ResourceQuota
+
+The design will be like:
+
+```go
+// ResourceQuotaSpec defines the desired hard limits to enforce for Quota
+type ResourceQuotaSpec struct {
+  // Hard is the set of desired hard limits for each named resource
+  // +optional
+  Hard ResourceList
+  // A collection of filters that must match each object tracked by a quota.
+  // If not specified, the quota matches all objects.
+  // +optional
+  Scopes []ResourceQuotaScope
+  // If specified, indicates this ResourceQuota applies to that priority class
+  // If not specified, this ResourceQuota applies to pods with no PriorityClassName.
+  // +optional
+  PriorityClassName string
+}
+```
+PriorityClassName in ResourceQuota is matched against PriorityClassName in the Pod, including the case where PriorityClassName is not set. In detail:
+
+* One could expect when `PriorityClassName` field is set, this ResourceQuota object matches pods with the same priority class.
+* If this field is not present, it indicates this quota matches to pods that do not have a PriorityClassName (Pods' integer priorities would be ignored).
+
+It is worth noting that quota is assigned per namespace, not per user, in Kubernetes. Adding priority class name to ResourceQuota will not change anything with respect to the scope of ResourceQuota. In other words, when priority class name is defined for a ResourceQuota object, the quota is applied to all pods at that priority class in the specified namespace.
+
+Unlike pod spec, ResourceQuota does not have Priority *int32 in its spec.
+As explained above, ResourceQuota with a PriorityClassName matches against pods with the same PriorityClassName, regarless of the integer value of priority.
+
+### Expected behavior of ResourceQuota admission controller
+
+This part is about how to make admission controller work with priority aware `ResourceQuota`.
+
+* ResourceQuota admission controller should reject creation or update of ResourceQuota object if new ResourceQuota names a PriorityClass that does not exist, i.e. is not a pre-defined or user-specified PriorityClass (but empty is allowed).
+* ResourceQuota admission controller should reject creation of ResourceQuota object if another ResourceQuota object with same PriorityClass already exists.
+* ResourceQuota admission controller should reject update of ResourceQuota object if the new (updated) PriorityClass matches that of another ResourceQuota object.
```

Comment:
could we make priority immutable post creation on quota?

Reply:
What if users just hope to update the priority class name? Or it's just a fake requirement?

Reply:
they could just as easily create a new quota and delete the old.  i am generally opposed to this trying to restrict against overlapping quotas, so i am pushing back on anything that attempts to protect against that.  a request for resource must satisfy all quotas, so there is no material harm by trying to protect against overlapping.  finally, i should be able to have quotas with matching priority but any permutation of scopes (besteffort, not best effort, terminating, not terminating).

Reply:
Yeah, so I can get your point here, I will change the statement for quota operations to avoid introducing extra complexity.